File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        rows = len(matrix)
        cols = len(matrix[0])
        ans = []
        left,right,up,down = 0, cols-1, 0, rows-1

        while len(ans)<rows*cols:
            for i in range(left, right+1):
                ans.append(matrix[up][i])
            up += 1
            for i in range(up, down+1):
                ans.append(matrix[i][right])
            right -= 1
            if up <= down:
                for i in range(right, left-1, -1):
                    ans.append(matrix[down][i])         
                down -= 1
            if left <= right:
                for i in range(down, up-1, -1):
                    ans.append(matrix[i][left])         
                left += 1                   

        return ans
'''
     0. 1. 2  3  4  5
    
0    1  2  3  4  5  6
1    18 19          7 
2    17             8
3    16             9
4    15 14 13 12 11 10


'''                

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        int left,right,up,down;
        vector<int> res;

        left = 0;
        right = matrix[0].size()-1;
        up = 0;
        down = matrix.size()-1;

        while(left<=right && up<=down) {
            int i=0;

            i=left;
            while(i<=right) {
                res.push_back(matrix[up][i++]);
            }
            i=++up;
            if(up>down) {
                break;
            }
            while(i<=down) {
                res.push_back(matrix[i++][right]);
            }
            i=--right;
            if(left>right) {
                break;
            }
            while(i>=left) {
                res.push_back(matrix[down][i--]);
            }
            i=--down;
            if(up>down) {
                break;
            }
            while(i>=up) {
                res.push_back(matrix[i--][left]);
            }
            left++;
        }

        return res;
    }
};

/*

 right - down 
  |       |
  up   - left

right - right max && visited
down - bottom max && visited
left - left min && visited
up - up min && visited



*/